### PR TITLE
IN-7 Create registration of user

### PR DESCRIPTION
**Pull request by glachezar**

### CODE REVIEW

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+{
+    readonly IMediator _mediator;
+
+    public UserController(IMediator mediator)
+    {
+        _mediator = mediator;
+    }
+
+    [HttpPost]
+    [AllowAnonymous]
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Controllers/UserController.cs:21]: Done.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Program.cs [hunk 2]
```diff
@@ -102,14 +102,18 @@
 
 #endregion Configure Application
 
+app.UseRouting();
 app.UseEndpoints(endpoints =>
 {
-    endpoints.MapDefaultControllerRoute();
+    //endpoints.MapDefaultControllerRoute();
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Program.cs:108]: Done.

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+using System.Text;
+using System.Text.Json;
+using Application.Commands;
+using MediatR;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+[Route("api/[controller]")]
+[ApiController]
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs:11]: Done.

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    public UserController(IMediator mediator)
+    {
+        _mediator = mediator;
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> CreateUser([FromBody] CreateUserCommand command)
+    {
+        var json = JsonSerializer.Serialize(command);
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs:25]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    public string Password { get; set; }
+
+    [Required]
+    [EmailAddress]
+    public string Email { get; set; }
+}
+
+internal class CreateUserCommandHandler : IRequestHandler<CreateUserCommand, object>
+{
+    public Task<object> Handle(CreateUserCommand request, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs:23]: This is changed with the moving of the logic from the controller to the command handler.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        _userManager = userManager;
+    }
+
+    public async Task<IdentityResult> Handle(RegisterUserCommand request, CancellationToken cancellationToken)
+    {
+        var user = new ApplicationUser
+        {
+            UserName = request.UserName,
+            Email = request.Email,
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs:28]: Added attributes to the command including email validation attribute.
Let me know if you had something else in mind.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Controllers/UserController.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    public async Task<IActionResult> CreateUser([FromBody]RegisterUserCommand command)
+    {
+        var result = await _mediator.Send(command);
+        if (result.Succeeded)
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Controllers/UserController.cs:25]: Unneccessery checks removed.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            //This might need to be changed in future.
+            EmailConfirmed = true,
+            AccountEnabled = true,
+        };
+
+        var result = await _userManager.CreateAsync(user, request.Password);
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs:39]: I have not try to login. I just checked if it actually imports the user to the database. 
I will need assistance to start the front end.

[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs:49]: I have tried to resolve the problem, but I was unable to do so.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs:44]: Done.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs]: All fixed it is from the login method in the account controller. You might want to have a look at it as the fix might not be a great quality code.

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            request.Email,
+        };
+
+        var json = JsonSerializer.Serialize(user);
+        var httpClient = new HttpClient();
+        var content = new StringContent(json, Encoding.UTF8, "application/json");
+        var response = await httpClient.PostAsync("https://localhost:5001/api/User", content);
+        var responseString = await response.Content.ReadAsStringAsync();
+
+        return responseString;
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs:40]: What should I return then?

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            request.Username,
+            request.Password,
+            request.Email,
+        };
+
+        var json = JsonSerializer.Serialize(user);
+        var httpClient = new HttpClient();
+        var content = new StringContent(json, Encoding.UTF8, "application/json");
+        var response = await httpClient.PostAsync("https://localhost:5001/api/User", content);
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs:37]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs:40]: Done.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs]: Done.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
+﻿namespace TRINV.IdentityServer.Commands;
+
+using System.ComponentModel.DataAnnotations;
+using MediatR;
+using Microsoft.AspNetCore.Identity;
+using TRINV.IdentityServer.Data.Models;
+
+public class RegisterUserCommand : IRequest<IdentityResult>
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/RegisterUserCommand.cs:12]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
+﻿namespace TRINV.Application.Commands;
+
+using System.ComponentModel.DataAnnotations;
+using System.Text;
+using System.Text.Json;
+using MediatR;
+
+public record CreateUserCommand : IRequest<string>
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateUserCommand.cs:8]: Done.

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+using MediatR;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+[AllowAnonymous]
+[ApiController]
+[Route("api/[controller]")]
+public class UserController : ControllerBase
+{
+    private readonly IMediator _mediator;
```
[the remark this answers is not in the document]
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs:15]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateIdentityUserCommand.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    [Required]
+    [StringLength(100, ErrorMessage = "The Password must be between 6 and 100 characters long.", MinimumLength = 6)]
+    public string Password { get; set; }
+
+    [Required]
+    [EmailAddress]
+    public string Email { get; set; }
+}
+
+public class CreateUserCommandHandler : IRequestHandler<CreateIdentityUserCommand>
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Commands/CreateIdentityUserCommand.cs:23]: The handler name should be CreateIdentityUserCommandHandler !

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+    public UserController(IMediator mediator)
+    {
+        _mediator = mediator;
+    }
+
+    [HttpPost]
+    public async Task CreateUser(CreateIdentityUserCommand command)
+    { 
+        await _mediator.Send(command);
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs:25]: Can't we return Ok();

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/CreateUserCommand.cs [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+public class CreateUserCommandHandler : IRequestHandler<CreateUserCommand, IdentityResult>
+{
+    readonly UserManager<ApplicationUser> _userManager;
+
+    public CreateUserCommandHandler(UserManager<ApplicationUser> userManager)
+    {
+        _userManager = userManager;
+
```
Comment (MihaelPavlov) [inline investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/CreateUserCommand.cs:28]: empty line

File: investment-app/TRINV.Api/TRINV.Application/Commands/CreateIdentityUserCommand.cs (hunk 15 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Commands/CreateIdentityUserCommand.cs:23]: Done.

File: investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/CreateUserCommand.cs (hunk 17 above)
Comment (glachezar) [reply investment-app/TRINV.IdentityServer/TRINV.IdentityServer/TRINV.IdentityServer/Commands/CreateUserCommand.cs:28]: Done.

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs (hunk 16 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/UserController.cs:25]: Done.

**Review by MihaelPavlov — APPROVED** [review]

**Pull request merged**